### PR TITLE
Add James's changes

### PR DESCRIPTION
**Pull request by JasonLovesDoggo**

### CODE REVIEW

**Comment by JasonLovesDoggo** [comment]
dont merge yet, pending review

File: core/models.py [hunk 1]
```diff
@@ -439,15 +467,51 @@ def save(self, *args, **kwargs):
         if self.end_date and timezone.is_naive(self.end_date):
             self.end_date = timezone.make_aware(self.end_date)
 
+        if self.review_status == ReviewStatus.Approved:
+            self.is_public = True
+        if self.review_status == ReviewStatus.Rejected:
```
Comment (JasonLovesDoggo) [inline core/models.py]: elif

File: core/models.py [hunk 2]
```diff
@@ -439,15 +467,51 @@ def save(self, *args, **kwargs):
...
+        )
+        if len(h) > 0 and h[0].id != self.id:
+            raise ValidationError(
+                f"An object with the name '{self.name}' already exists."
+            )
         super().save(*args, **kwargs)
 
 
+class Vote(models.Model):
+    type_vote = models.BooleanField(default=True)
```
Comment (JasonLovesDoggo) [inline core/models.py]: change to positive vs "vote type"

File: core/models.py [hunk 3]
```diff
@@ -439,15 +467,51 @@ def save(self, *args, **kwargs):
...
+    hackathon_id = models.OneToOneField(
```
Comment (JasonLovesDoggo) [inline core/models.py]: should just be hackathon not hackathon_id

File: core/models.py [hunk 4]
```diff
@@ -439,15 +467,51 @@ def save(self, *args, **kwargs):
...
+        Hackathon, on_delete=models.CASCADE, related_name="votes"
+    )
+    from_hacker = models.ManyToManyField(Hacker, related_name="hackathon_votes")
```
Comment (JasonLovesDoggo) [inline core/models.py]: should just be called "hacker"

File: core/models.py [hunk 5]
```diff
@@ -439,15 +467,51 @@ def save(self, *args, **kwargs):
...
+        return f"Hacker(type_vote={self.type_vote}, hackathon_id={self.hackathon_id}, from_hacker={self.from_hacker}, time_created={self.time_created})"
+
+
 class CuratorRequest(models.Model):
     hackathon = models.ForeignKey(Hackathon, on_delete=models.CASCADE)
     team_name = models.CharField(max_length=255)
     team_description = models.TextField()
     reason = models.TextField()
     created_at = models.DateTimeField(auto_now_add=True)
+    created_by = models.ManyToManyField(Hacker, related_name="curator_requests")
```
Comment (JasonLovesDoggo) [inline core/models.py]: "curation_requests" is better

File: core/views.py [hunk 6]
```diff
@@ -45,69 +59,90 @@ def home(request):
     return render(request, "home.html", context)
 
 
-@ratelimit(key="user_or_ip", rate="1/m", block=True)
```
Comment (JasonLovesDoggo) [inline core/views.py:48]: readd

File: core/views.py [hunk 7]
```diff
@@ -45,69 +59,90 @@ def home(request):
...
+# @ratelimit(key="user_or_ip", rate="1/m", block=True)
 def addHackathons(request):
+    print(request.method)
```
Comment (JasonLovesDoggo) [inline core/views.py]: remove debug prints

File: core/views.py [hunk 8]
```diff
@@ -45,69 +59,90 @@ def home(request):
...
+        print("got the post requets")
         if form.is_valid():
+            print("sending the form")
+            form.save(commit=False)
+            form.review_status = ReviewStatus.Pending
+            form.created_by = Hacker.objects.get(id=request.user.id)
+            form.save()
             return redirect("home")
+        else:
+            print(form.errors)
```
Comment (JasonLovesDoggo) [inline core/views.py]: add error handeling

File: core/views.py [hunk 9]
```diff
@@ -45,69 +59,90 @@ def home(request):
...
-    template_name = "hackathons/hackathons.html"
-    context_object_name = "hackathons"
-    paginate_by = 30
+    user_votes = Vote.objects.filter(from_hacker=request.user.id)
+    user_votes_dict = {
+        vote.hackathon_id.duplication_id: vote.type_vote for vote in user_votes
+    }
+    print(user_votes_dict)
+    query_base = Q(end_date__gte=tdy_date, is_public=True)
+    if country and country != "none" and country != "World":
```
Comment (JasonLovesDoggo) [inline core/views.py:91]: much better ways to do this but if it works it passess

File: core/views.py [hunk 10]
```diff
@@ -148,50 +183,77 @@ def setting(request):
...
-            }
-        )
 
+def str_to_bool(s):
+    return s.lower() in ["true", "1", "yes"]
 
-@login_required
-def unsave_hackathon(request: HttpRequest, hackathon_id):
+
+def add_vote(request: HttpRequest, hackathon_id):
```
Comment (JasonLovesDoggo) [inline core/views.py]: should have login required

File: core/views.py [hunk 11]
```diff
@@ -148,50 +183,77 @@ def setting(request):
...
     if request.method == "POST":
         hackathon = get_object_or_404(Hackathon, id=hackathon_id)
         hacker = get_object_or_404(Hacker, id=request.user.id)
+        type_vote = request.GET.get("vote_state")
+        state = str_to_bool(type_vote)
+        vote = Vote.objects.filter(hackathon_id=hackathon, from_hacker=hacker)
+
+        if vote.exists():
```
Comment (JasonLovesDoggo) [inline core/views.py]: checking if .exists then .first causes an n+1 query. just run .first and check if None

File: core/views.py [hunk 12]
```diff
@@ -148,50 +183,77 @@ def setting(request):
...
+                hackathon.count_downvotes -= 1
+                hackathon.count_upvotes += 1
 
-        if not hacker.saved.filter(id=hackathon_id).exists():
-            return JsonResponse(
-                {"status": "fail", "error": "Invalid request"}, status=400
-            )
+            # Update the vote type
+            vote.type_vote = state
+            vote.save()
```
Comment (JasonLovesDoggo) [inline core/views.py]: a simple vote.update() would be shorter and cleaner

File: core/views.py [hunk 13]
```diff
@@ -148,50 +183,77 @@ def setting(request):
...
-                "hackathon": {
-                    "id": hackathon.id,
-                    "name": hackathon.name,
-                },
-            }
-        )
+            vote.from_hacker.add(hacker)
+
+        hackathon.count_upvotes = Vote.objects.filter(
+            hackathon_id=hackathon, type_vote=True
```
Comment (JasonLovesDoggo) [inline core/views.py]: You'll want a votes field on the Hackathon model as this method is super slow  and non scaleable

File: core/views.py [hunk 14]
```diff
@@ -226,16 +289,58 @@ class SavedHackathonsPage(ListView):
...
+@login_required
+@user_passes_test(is_admin)
+@ratelimit(key="user_or_ip", rate="1/d", block=True)
 def scrape(request):
     scrape_all()
     return HttpResponse("Scraped!")
+
+
+@cache_page(60 * 60)  # 1 hour cache
+def calednar_genator(request):
```
Comment (JasonLovesDoggo) [inline core/views.py]: typos

File: static/css/styles.css [hunk 15]
```diff
@@ -308,21 +308,25 @@ a {
...
 .flip-card-front,
 .flip-card-back {
   position: absolute;
   backface-visibility: hidden;
 }
 
-.flip-card-back {
-  transform: rotateY(180deg);
+/*.flip-card-back {*/
+/*  transform: rotateY(180deg);*/
```
Comment (JasonLovesDoggo) [inline static/css/styles.css]: remove dead code

File: static/js/hackathons.js [hunk 16]
```diff
@@ -1,9 +1,49 @@
...
+
+    const upButton = document.getElementById(`${hackathonId}-up`);
+    const downButton = document.getElementById(`${hackathonId}-down`);
+    const upText = document.getElementById(`${hackathonId}-up-text`);
+    const downText = document.getElementById(`${hackathonId}-down-text`);
+
+
+    if (vote_state === 'true') {
+        // Check if the down button is active
+        if (downButton.classList.contains('fill-black')) {
```
Comment (JasonLovesDoggo) [inline static/js/hackathons.js:55]: this code is useless as it's overwritten in the next codeblock

File: static/js/hackathons.js [hunk 17]
```diff
@@ -1,9 +1,49 @@
...
+        }
+
+        // Activate the up button
+        upButton.classList.add('fill-black');
+        upButton.classList.remove('fill-white');
+        upText.textContent = parseInt(upText.textContent) + 1;
+    } else {
+        // Check if the up button is active
+        if (upButton.classList.contains('fill-black')) {
+            upButton.classList.remove('fill-black');
```
Comment (JasonLovesDoggo) [inline static/js/hackathons.js:77]: same here

File: core/views.py (hunk 6 above)
Comment (JasonLovesDoggo) [reply core/views.py:48]: Re add

File: core/models.py (hunk 5 above)
Comment (jamesjamcow) [reply core/models.py]: what should I change to curation_requests, I assume it was: created_by = models.ManyToManyField(Hacker, related_name="curation_requests")

File: core/models.py [hunk 18]
```diff
@@ -439,15 +467,51 @@ def save(self, *args, **kwargs):
...
+
+        h = Hackathon.objects.filter(
+            name=self.name,
+            end_date__year=self.end_date.year,
+            end_date__month=self.end_date.month,
+        )
+        if len(h) > 0 and h[0].id != self.id:
+            raise ValidationError(
+                f"An object with the name '{self.name}' already exists."
+            )
```
Comment (ThePathOfLeastResistance) [inline core/models.py:483]: why not just so dup id ? its prob faster than query the db
```suggestion
        h = Hackathon.objects.filter(
            name=self.name,
            end_date__year=self.end_date.year,
            end_date__month=self.end_date.month,
        )
        if len(h) > 0 and h[0].id != self.id:
            raise ValidationError(
                f"An object with the name '{self.name}' already exists."
            )
```

Comment (JasonLovesDoggo) [inline core/models.py]: James... it's your code

File: core/models.py [hunk 19]
```diff
@@ -439,15 +467,51 @@ def save(self, *args, **kwargs):
         if self.end_date and timezone.is_naive(self.end_date):
             self.end_date = timezone.make_aware(self.end_date)
 
+        if self.review_status == ReviewStatus.Approved:
+            self.is_public = True
+        if self.review_status == ReviewStatus.Rejected:
+            self.is_public = False
+
+        h = Hackathon.objects.filter(
```
Comment (JasonLovesDoggo) [inline core/models.py]: please use better var names

File: core/models.py [hunk 20]
```diff
@@ -439,15 +467,51 @@ def save(self, *args, **kwargs):
...
+            name=self.name,
+            end_date__year=self.end_date.year,
+            end_date__month=self.end_date.month,
```
Comment (JasonLovesDoggo) [inline core/models.py]: add id=self.id then check h.exists()

File: core/models.py (hunk 18 above)
Comment (jamesjamcow) [reply core/models.py]: that's amol's code

**Comment by JasonLovesDoggo** [comment]
Also, before a merge is approved you will have to delete all of your new migrations and recreate them in one.

File: core/models.py (hunk 18 above)
Comment (JasonLovesDoggo) [reply core/models.py]: Then don't comment about it on your PR. amol has one open

File: core/views.py [hunk 21]
```diff
@@ -45,69 +59,87 @@ def home(request):
...
     if request.method == "POST":
         form = HackathonForm(request.POST)
         if form.is_valid():
+            print("sending the form")
+            form.save(commit=False)
+            form.review_status = ReviewStatus.Pending
+            form.created_by = Hacker.objects.get(id=request.user.id)
+            form.save()
             return redirect("home")
+        else:
```
Comment (jamesjamcow) [inline core/views.py]: add popups telling the user if the form was submited successfully or rejected

Comment (JasonLovesDoggo) [reply core/views.py]: see form.errors https://docs.djangoproject.com/en/5.1/ref/forms/api/#django.forms.Form.errors

**Comment by jamesjamcow** [comment]
Everything done for V1

File: core/filters.py [hunk 22]
```diff
@@ -0,0 +1,55 @@
+import django_filters
+from .models import Hackathon
+from django.db.models import Q
+from django.utils import timezone
+
+
+# filters
```
Comment (JasonLovesDoggo) [inline core/filters.py]: Not really a useful comment (doesn't impact anything)

File: core/filters.py [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+
+    def filter_by_proximity(self, queryset, name, value):
+        # Get latitude and longitude from the query parameters
+        lat = self.request.GET.get("latitude")
+        lon = self.request.GET.get("longitude")
+        if lat and lon:
+            lat, lon = float(lat), float(lon)
+            queryset = queryset.filter(
+                Q(
+                    latitude__range=(lat - value, lat + value),
```
Comment (JasonLovesDoggo) [inline core/filters.py:50]: should be done using geodjango in the future (just noting it down)

File: core/forms.py [hunk 24]
```diff
@@ -8,26 +9,29 @@
...
 
 class HackathonForm(forms.ModelForm):
-    short_name = forms.CharField(max_length=255)
+    short_name = forms.CharField(max_length=255, required=False)
     name = forms.CharField(max_length=255)
     website = forms.URLField()
     country = forms.CharField(max_length=255)
     city = forms.CharField(max_length=255)
-    image = forms.ImageField()
+    # image = forms.ImageField()
```
Comment (JasonLovesDoggo) [inline core/forms.py]: how come no image?

File: core/forms.py [hunk 25]
```diff
@@ -53,6 +57,13 @@ class Meta:
             "is_public",
             "notes",
             "metadata",
+            "categories",
+            "created_by",
+            "hybrid",
+            "count_upvotes",
+            "notification_policy",
+            "count_downvotes",
+            "image",
```
Comment (JasonLovesDoggo) [inline core/forms.py]: you just removed the image field?

File: core/models.py [hunk 26]
```diff
@@ -260,20 +268,35 @@ class Location(models.Model):
...
-    longitude = DecimalField(max_digits=22, decimal_places=16)
+    latitude = DecimalField(max_digits=22, decimal_places=16, blank=True, null=True)
+    longitude = DecimalField(max_digits=22, decimal_places=16, blank=True, null=True)
 
 
 class HackathonLocation(models.Model):
     name = models.CharField(
         max_length=255,
-        help_text="Where the hackathon is located (e.g. University of Toronto)",
+        help_text="Where the hackathon is located (e.g. Buringlont, Ontario)",
```
Comment (JasonLovesDoggo) [inline core/models.py:288]: Buringlont :(

File: core/models.py [hunk 27]
```diff
@@ -388,33 +419,33 @@ class Hackathon(MetaDataMixin):
...
+    )
+    is_nonenglish = models.BooleanField(
+        default=False, help_text="Is the primary language of this hackathon not English"
+    )
+
     custom_info = models.JSONField(
         default=dict, null=True, blank=True
     )  # Anything else that we might want to add in a structured format
 
+    vote_count = models.BigIntegerField(default=0)
```
Comment (JasonLovesDoggo) [inline core/models.py]: hmm... I would rename to rating or something like net_vote as it's not really a count but a sum of the negative and positive votes

File: core/models.py [hunk 28]
```diff
@@ -436,18 +467,52 @@ def __str__(self):
...
 
 
+class Vote(models.Model):
+    vote_type = models.BooleanField(default=True)
+    hackathon = models.OneToOneField(
+        Hackathon, on_delete=models.CASCADE, related_name="votes"
+    )
+    hacker = models.ManyToManyField(Hacker, related_name="hackathon_votes")
+    time_created = models.DateTimeField(auto_now_add=True)
+    time_updated = models.DateTimeField(auto_now=True)
```
Comment (JasonLovesDoggo) [inline core/models.py]: You can't update a vote

File: core/models.py [hunk 29]
```diff
@@ -436,18 +467,52 @@ def __str__(self):
...
+                " ", ""
+            ) + self.end_date.strftime("-%Y")
+        else:
+            pass
+
         super().save(*args, **kwargs)
 
 
+class Vote(models.Model):
+    vote_type = models.BooleanField(default=True)
```
Comment (JasonLovesDoggo) [inline core/models.py]: what does "True" mean. clarify naming to indicate meaning. e.g. positive or `is_upvote`

File: core/models.py [hunk 30]
```diff
@@ -436,18 +467,52 @@ def __str__(self):
...
     created_at = models.DateTimeField(auto_now_add=True)
+    created_by = models.ManyToManyField(Hacker, related_name="curation_requests")
+    review_status = models.CharField(
+        max_length=255,
+        blank=True,
+        null=False,
+        default=ReviewStatus.Pending,
+        help_text="Status of the review process",
+        choices=ReviewStatus.choices,
+    )
```
Comment (JasonLovesDoggo) [inline core/models.py:543]: I would add a Reviewed_by field (staff member) could be only used in v2 though

File: core/urls.py [hunk 31]
```diff
@@ -30,9 +29,10 @@
         name="save_hackathon",
     ),
     path(
-        "hackathons/<int:hackathon_id>/unsave",
-        views.unsave_hackathon,
-        name="unsave_hackathon",
+        "hackathons/<int:hackathon_id>/vote",
+        views.add_vote,
+        name="save_hackathon",
```
Comment (JasonLovesDoggo) [inline core/urls.py]: rename to  `hackathon_vote` or `vote_hackathon`

File: core/urls.py [hunk 32]
```diff
@@ -30,9 +29,10 @@
...
     ),
     path("scrape/", views.scrape, name="scrape"),
+    path("export_cal/", views.calendar_generator, name="calednar_genator"),
```
Comment (JasonLovesDoggo) [inline core/urls.py]: *calendar_generator

File: core/views.py [hunk 33]
```diff
@@ -1,31 +1,44 @@
...
 from django.shortcuts import render
 from django.views.generic import ListView
-from core.models import Hackathon, Hacker
+from django.shortcuts import redirect
+from django.apps import apps
+from django.db.models import Q
+from icalendar import Calendar, Event
+from django_ratelimit.decorators import ratelimit
+
+if apps.ready:
```
Comment (JasonLovesDoggo) [inline core/views.py]: why is this added?

File: core/views.py (hunk 6 above)
Comment (JasonLovesDoggo) [reply core/views.py:48]: add back

File: core/views.py (hunk 21 above)
Comment (JasonLovesDoggo) [reply core/views.py]: this was not changed

File: core/views.py [hunk 34]
```diff
@@ -153,45 +206,83 @@ def save_hackathon(request: HttpRequest, hackathon_id):
...
-def unsave_hackathon(request: HttpRequest, hackathon_id):
+def add_vote(request: HttpRequest, hackathon_id):
     if request.method == "POST":
         hackathon = get_object_or_404(Hackathon, id=hackathon_id)
         hacker = get_object_or_404(Hacker, id=request.user.id)
+        type_vote = request.GET.get("vote_state")
+        state = str_to_bool(type_vote)
+        vote = Vote.objects.filter(hackathon=hackathon, hacker=hacker).first()
+
+        if vote is not None:
```
Comment (JasonLovesDoggo) [inline core/views.py]: 1. remove debug prints
2. use hackathons.update() with a F() + 1 or -1 instead of += 1
3. exit early instead of adding additional nesting (everything is inside of a `if vote is not None`

File: core/views.py [hunk 35]
```diff
@@ -226,16 +312,57 @@ class SavedHackathonsPage(ListView):
 
     def get_queryset(self):
         user: Hacker = self.request.user
-        return user.saved.all()
+        saved_hackathons = user.saved.all()
```
Comment (JasonLovesDoggo) [inline core/views.py]: inline variable

File: core/views.py [hunk 36]
```diff
@@ -226,16 +312,57 @@ class SavedHackathonsPage(ListView):
...
+        return saved_hackathons
 
 
 # checks if the user is an admin
-def is_admin(user):
+def is_admin(user: AbstractUser | AnonymousUser):
     return user.is_superuser
 
 
+# @login_required
```
Comment (JasonLovesDoggo) [inline core/views.py]: add back auth to this. (only admins + only logged in)

File: static/js/hackathons.js [hunk 37]
```diff
@@ -1,9 +1,100 @@
...
+    const popup = document.getElementById('login-popup');
+    popup.classList.remove('hidden');
+    popup.classList.add('flex');
+}
+
+
+function voteHackathon(event, hackathonId, vote_state) {
+    event.preventDefault();
+    var xhr = new XMLHttpRequest();
+
```
Comment (JasonLovesDoggo) [inline static/js/hackathons.js:47]: do you need a CSRF token here?

**Review by JasonLovesDoggo — COMMENTED** [review]
(requesting changes but it won't let me click request changes)

File: core/forms.py (hunk 24 above)
Comment (jamesjamcow) [reply core/forms.py]: before it wasnt working and idk what was wrong, I added it back and it seems to work

File: core/models.py (hunk 28 above)
Comment (jamesjamcow) [reply core/models.py]: like if they change it from a upvote to a downvote, is this field even needed?

Comment (jamesjamcow) [reply core/models.py]: I just deleted  time_updated = models.DateTimeField(auto_now=True) for now

File: core/views.py (hunk 33 above)
Comment (jamesjamcow) [reply core/views.py]: this is for migration, when I first added this and made migrations, it would import it while doing migrations or smt like that. Ill delete it

File: core/forms.py (hunk 24 above)
Comment (JasonLovesDoggo) [reply core/forms.py]: If you reference the image in the field below you can't just comment it out

File: core/models.py (hunk 28 above)
Comment (JasonLovesDoggo) [reply core/models.py]: > like if they change it from a upvote to a downvote, is this field even needed?

no. We don't care about that data

Comment (JasonLovesDoggo) [reply core/models.py]: you could update time_created in that case

File: core/views.py (hunk 33 above)
Comment (JasonLovesDoggo) [reply core/views.py]: ah, fair enough.

File: core/views.py (hunk 8 above)
Comment (jamesjamcow) [reply core/views.py]: I am not quite sure how to do this, like I get the error so do I return it as a http response ?

File: static/js/hackathons.js (hunk 37 above)
Comment (jamesjamcow) [reply static/js/hackathons.js:47]: yea, I have it at the bottem with xhr.setRequestHeader('X-CSRFToken', getCookie('csrftoken'));

Comment (JasonLovesDoggo) [reply static/js/hackathons.js:47]: Missed that

File: core/views.py (hunk 9 above)
Comment (jamesjamcow) [reply core/views.py:91]: we will def have a better system for v2

File: core/views.py (hunk 34 above)
Comment (jamesjamcow) [reply core/views.py]: what does "exit early instead of adding additional nesting (everything is inside of a if vote is not None" mean?

**Comment by JasonLovesDoggo** [comment]
LGTM! https://github.com/Hackathons-Canada/HackathonsCanada/issues/21 is an issue but it's non blocking
